tests/test_api.rs

```diff
@@ -7078,3 +7078,170 @@ fn test_fast_calls() {
   eval(scope, source).unwrap();
   assert_eq!("fast", unsafe { WHO });
 }
+
+#[test]
+fn test_fast_calls_sequence() {
+  static mut WHO: &str = "none";
+  fn fast_fn(
+    _recv: v8::Local<v8::Object>,
+    a: u32,
+    b: u32,
+    array: v8::Local<v8::Array>,
+  ) -> u32 {
+    unsafe { WHO = "fast" };
+    assert_eq!(array.length(), 2);
+    a + b + array.length()
+  }
+
+  pub struct FastTest;
+  impl fast_api::FastFunction for FastTest {
+    fn args(&self) -> &'static [fast_api::Type] {
+      &[
+        fast_api::Type::V8Value,
+        fast_api::Type::Uint32,
+        fast_api::Type::Uint32,
+        fast_api::Type::Sequence(fast_api::CType::Void),
+      ]
+    }
+
+    fn return_type(&self) -> fast_api::CType {
+      fast_api::CType::Uint32
+    }
+
+    type Signature = fn(
+      receiver: v8::Local<v8::Object>,
+      a: u32,
+      b: u32,
+      array: v8::Local<v8::Array>,
+    ) -> u32;
+    fn function(&self) -> Self::Signature {
+      fast_fn
+    }
+  }
+
+  fn slow_fn(
+    scope: &mut v8::HandleScope,
+    _: v8::FunctionCallbackArguments,
+    mut rv: v8::ReturnValue,
+  ) {
+    unsafe { WHO = "slow" };
+    rv.set(v8::Boolean::new(scope, false).into());
+  }
+
+  let _setup_guard = setup();
+  let isolate = &mut v8::Isolate::new(Default::default());
+  let scope = &mut v8::HandleScope::new(isolate);
+  let context = v8::Context::new(scope);
+  let scope = &mut v8::ContextScope::new(scope, context);
+
+  let global = context.global(scope);
+
+  let template =
+    v8::FunctionTemplate::builder(slow_fn).build_fast(scope, FastTest);
+
+  let name = v8::String::new(scope, "func").unwrap();
+  let value = template.get_function(scope).unwrap();
+  global.set(scope, name.into(), value.into()).unwrap();
+  let source = r#"
+  function f(x, y, data) { return func(x, y, data); }
+  %PrepareFunctionForOptimization(f);
+  const arr = [3, 4];
+  f(1, 2, arr);
+"#;
+  eval(scope, source).unwrap();
+  assert_eq!("slow", unsafe { WHO });
+
+  let source = r#"
+    %OptimizeFunctionOnNextCall(f);
+    f(1, 2, arr);
+  "#;
+  eval(scope, source).unwrap();
+  assert_eq!("fast", unsafe { WHO });
+}
+
+#[repr(C)]
+pub struct FastApiArrayBuffer {
+  byte_length: usize,
+  data: *mut u32,
+}
+
+#[test]
+fn test_fast_calls_arraybuffer() {
+  static mut WHO: &str = "none";
+  fn fast_fn(
+    _recv: v8::Local<v8::Object>,
```

Comment:
Is the `recv` parameter required in Fast API calls? I somehow thought that is not the case and that it's controlled by an option in the CFunctionInfo or something?

Reply:
Reading the sources a bit, I'm definitely wrong. Seems like the receiver object is a required parameter, and only the options parameter is optional.

Reply:
It's lovely to have good documentation, innit...

Good to hear that you managed to cut through the debugging. Also, somewhat of a bummer to hear that typed arrays cannot be passed without this parameter. But then again, as long as it's a V8 TypedArray object and not a pointer to linear memory that the Fast API Call gets, it doesn't matter what sort of extra parameters are required since any FFI call (which is of course what I'm kind of thinking about) will require an intermediate function anyway to map from the TypedArray to the pointer.

But, I presume this might be entirely acceptable, even preferable, for Fast API op calls since you can create a custom receiver object that carries the `Rc<RefCell<JsRuntimeState>>` to enable essentially equivalent internal works as normal op calls. So I must say, the V8 did do their homework with thinking up this API. All hail Google.